admin/client/Forms.tsx

```diff
@@ -157,13 +157,40 @@ export interface NumberFieldProps {
     helpText?: string
 }
 
-export class NumberField extends React.Component<NumberFieldProps> {
+/**
+ * Like NumberField, but also capture the users text input
+ */
+export class NumberInputField extends React.Component<
+    NumberFieldProps<NumericInputValue>
+> {
+    render() {
+        const { props } = this
+
+        const textFieldProps = extend({}, props, {
+            value: props.value && props.value ? props.value.input : undefined,
+            onValue: (value: string) => {
+                const asNumber = parseFloat(value)
+                const nValue: NumericInputValue = {
+                    input: value,
+                    value: isNaN(asNumber) ? undefined : asNumber
+                }
+                props.onValue(nValue)
+            }
+        })
+
+        return <TextField {...textFieldProps} />
+    }
+}
+
+export class NumberField extends React.Component<
+    NumberFieldProps<number | undefined>
+> {
     render() {
         const { props } = this
 
         const textFieldProps = extend({}, props, {
             value:
-                props.value !== undefined ? props.value.toString() : undefined,
+                props.value && props.value ? props.value.toString() : undefined,
```

Comment:
This line does `props.value && props.value` again, I must be missing something, what's going on with this?

Reply:
In any case, whatever logic is on here might make sense as a method on the NumericInputValue class? Then it could be given a meaningful method name, and wouldn't have to be repeated between the two field components

Reply:
The logic here is just to prevent calling a method on undefined. 
I'll update both places to use optional chaining e.g. `props.value?.toString()`. I didn't realize they had added that to typescript.